### PR TITLE
Dec proc lengths and preprocessing

### PR DESCRIPTION
**Pull request by vhavlena**
This PR collects lengths from the output of the decision procedure and applies preprocessing.

### CODE REVIEW

**Review by Adda0 — APPROVED** [review]
After a quick review, the changes seem sound. There will be some space for further optimization, but I would leave this for later.

File: src/smt/theory_str_noodler/decision_procedure.cpp [hunk 1]
```diff
@@ -353,14 +350,108 @@ namespace smt::noodler {
...
+        for(const BasicTerm& var : this->init_length_sensitive_vars) {
+            std::set<std::pair<int, int>> aut_constr = Mata::Strings::get_word_lengths(*ass.at(var));
+
+            auto it = variable_map.find(var);
+            expr_ref var_expr(this->m);
+            if(it != variable_map.end()) { // if the variable is not found, it was introduced in the preprocessing -> create a new z3 variable
+                var_expr = it->second;
+            } else {
+                var_expr = util::mk_str_var(var.get_name(), this->m, this->m_util_s);
+            }
```
[the remark this answers is not in the document]
Comment (vhavlena) [reply src/smt/theory_str_noodler/decision_procedure.cpp:372]: ok, I fix it

**Comment by vhavlena** [comment]
@Adda0  Can I merge it now?

**Comment by Adda0** [comment]
I would go for it.

**Pull request merged**